### PR TITLE
Add redirection awareness for RENAME table operation

### PR DESCRIPTION
**Pull request by findinpath**
## Description

Add redirection awareness for the `ALTER TABLE xxx RENAME TO yyy` statement.

The implementation follows the specification from https://github.com/trinodb/trino/issues/11202#issue-1150518940

> Is this change a fix, improvement, new feature, refactoring, or other?
This is a new feature.

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

This implementation affects the core query engine.

> How would you describe this change to a non-technical end user or system administrator?

In the context of working in an environment where Hive/Iceberg/Delta tables are mixed, allow the user to do a table rename by deriving the destination catalog and schema from the source table name.

When having the table redirection enabled in the `hive` connector (via `hive.iceberg-catalog-name=iceberg` in `hive.properties`) and a table `iceberg.default.iceberg_table_name` within the `iceberg` connector, any of the following statements is considered valid:

```
     ALTER TABLE hive.default.iceberg_table_name to iceberg_table_name_new;
     ALTER TABLE hive.default.iceberg_table_name to default.iceberg_table_name_new;
     ALTER TABLE hive.default.iceberg_table_name to iceberg.default.iceberg_table_name_new;
```

## Documentation

(x) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
(x) Release notes entries required with the following suggested text:

```markdown
# Main
* Add redirection awareness for RENAME table operation
```

### CODE REVIEW

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveRedirectionToIceberg.java [hunk 1]
```diff
@@ -305,11 +305,18 @@ public void testAlterTableRename()
...
         createIcebergTable(icebergTableName, false);
 
-        //TODO restore test assertions after adding redirection awareness to the RenameTableTask
-        assertQueryFailure(() -> onTrino().executeQuery("ALTER TABLE " + hiveTableName + " RENAME TO " + tableName + "_new"))
-                .hasMessageMatching("\\QQuery failed (#\\E\\S+\\Q): Cannot query Iceberg table 'default." + tableName + "'");
+        assertQueryFailure(() -> onTrino().executeQuery("ALTER TABLE " + hiveTableName + " RENAME TO hive.default." + tableName + "_new"))
+                .hasMessageMatching("\\QQuery failed (#\\E\\S+\\Q): line 1:1: Table rename across catalogs is not supported");
 
-        onTrino().executeQuery("DROP TABLE " + icebergTableName);
+        onTrino().executeQuery("ALTER TABLE " + hiveTableName + " RENAME TO " + tableName + "_new_without_catalog_without_schema");
```
Comment (findepi) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveRedirectionToIceberg.java:311]: define a `newTableName` variable

File: core/trino-main/src/main/java/io/trino/execution/RenameTableTask.java [hunk 2]
```diff
@@ -86,28 +95,44 @@ public String getName()
...
         }
         if (metadata.getTableHandle(session, target).isPresent()) {
             throw semanticException(TABLE_ALREADY_EXISTS, statement, "Target table '%s' already exists", target);
         }
-        if (!tableName.getCatalogName().equals(target.getCatalogName())) {
+        if (!tableHandle.getCatalogName().getCatalogName().equals(target.getCatalogName())) {
             throw semanticException(NOT_SUPPORTED, statement, "Table rename across catalogs is not supported");
         }
-        accessControl.checkCanRenameTable(session.toSecurityContext(), tableName, target);
+        accessControl.checkCanRenameTable(session.toSecurityContext(), tableName, source);
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/execution/RenameTableTask.java:117]: why target -> source ?

File: core/trino-main/src/main/java/io/trino/execution/RenameTableTask.java [hunk 3]
```diff
@@ -86,28 +95,44 @@ public String getName()
...
             if (!statement.isExists()) {
                 throw semanticException(TABLE_NOT_FOUND, statement, "Table '%s' does not exist", tableName);
             }
             return immediateVoidFuture();
         }
-
-        QualifiedObjectName target = createQualifiedObjectName(session, statement, statement.getTarget());
+        TableHandle tableHandle = redirectionAwareTableHandle.getTableHandle().get();
+        QualifiedObjectName source = redirectionAwareTableHandle.getRedirectedTableName().orElse(createQualifiedObjectName(session, statement, statement.getSource()));
+        QualifiedObjectName target = createTargetQualifiedObjectName(source, statement.getTarget());
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/execution/RenameTableTask.java:107]: This is a change in behavior that spans non-redirected use-cases.
For example
```
USE hive.default;
ALTER TABLE hive.tmp.my_table RENAME TO my_renamed;
```

before the change, the table was being renamed to `hive.default.my_renamed`.
after the change, the table is `hive.tmp.my_renamed`.
please correct me if am wrong.

I wasn't aware of this behavior. and this is why i wanted RENAME to work with redirects.
I actually think that the old behavior is misleading, hence https://github.com/trinodb/trino/issues/11281.
The changes here seem to resolve that problem, so

- let's make sure it's intentional (ie we agree https://github.com/trinodb/trino/issues/11281 is a bug)
- let's make this as a separate commit (or even PR)

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveRedirectionToIceberg.java [hunk 4]
```diff
@@ -305,11 +305,18 @@ public void testAlterTableRename()
...
+        onTrino().executeQuery("ALTER TABLE hive.default." + tableName + "_new_without_catalog_without_schema RENAME TO default." + tableName + "_new_without_catalog_with_schema");
+        onTrino().executeQuery("ALTER TABLE hive.default." + tableName + "_new_without_catalog_with_schema RENAME TO iceberg.default." + tableName + "_new_with_catalog_with_schema");
+
+        assertResultsEqual(
+                onTrino().executeQuery("TABLE " + icebergTableName + "_new_with_catalog_with_schema"),
+                onTrino().executeQuery("TABLE " + hiveTableName + "_new_with_catalog_with_schema"));
+
+        onTrino().executeQuery("DROP TABLE " + icebergTableName + "_new_with_catalog_with_schema");
```
Comment (findepi) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveRedirectionToIceberg.java:319]: does TestIcebergRedirectionToHive need an update too?

**Review by findepi — CHANGES_REQUESTED** [review]

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveRedirectionToIceberg.java (hunk 4 above)
Comment (findinpath) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveRedirectionToIceberg.java:322]: There exists already a test `io.trino.tests.product.iceberg.TestIcebergRedirectionToHive#testAlterTableRename`.

File: core/trino-main/src/main/java/io/trino/execution/RenameTableTask.java (hunk 3 above)
Comment (findinpath) [reply core/trino-main/src/main/java/io/trino/execution/RenameTableTask.java:103]: I'll create a new PR for this purpose.

File: core/trino-main/src/main/java/io/trino/execution/RenameTableTask.java (hunk 2 above)
Comment (findepi) [reply core/trino-main/src/main/java/io/trino/execution/RenameTableTask.java]: do we have test coverage for this?

Comment (findinpath) [reply core/trino-main/src/main/java/io/trino/execution/RenameTableTask.java]: We have a few tests:
- io.trino.security.TestAccessControl#testNonQueryAccessControl -> is using a rather naive io.trino.testing.TestingAccessControlManager behind the scenes (and no `systemAccessControls`)
- io.trino.testing.BaseConnectorTest#testRenameTable 

However I haven't found a test interacting with a fine grained system access control implementation like https://trino.io/docs/current/security/file-system-access-control.html

Comment (findepi) [reply core/trino-main/src/main/java/io/trino/execution/RenameTableTask.java]: Did any test fail when you had `target` here?

Comment (findinpath) [reply core/trino-main/src/main/java/io/trino/execution/RenameTableTask.java]: No. There were no failures as far as I can recall.
https://github.com/trinodb/trino/actions/runs/1922889046

Comment (findepi) [reply core/trino-main/src/main/java/io/trino/execution/RenameTableTask.java]: Can you please add a test? Orthogonal to this PR

Comment (findepi) [reply core/trino-main/src/main/java/io/trino/execution/RenameTableTask.java]: actually…. there should be two tests: we should have an access control test for rename; and we should have an access control test for rename with redirects; we should verify which table AC is asked about

Comment (findinpath) [reply core/trino-main/src/main/java/io/trino/execution/RenameTableTask.java]: Created https://github.com/trinodb/trino/pull/11488 as a scaffolding for dealing with access control tests for redirections. I'll be adding support for `RENAME TABLE` there once it reaches `master`.

**Review by findepi — APPROVED** [review]

**Pull request merged**